File: 369.rb
class ListNode
  attr_accessor :val, :next
  def initialize(val)
    @val = val
    @next = nil
  end
end

def plus_one(head)
  def plus(head)
    if head.nil?
      1
    else
      head.val = (head.val + plus(head.next)) % 10
      if head.val == 0
        1
      else
        0
      end
    end
  end
  plus(head)
  head
end

File: 369.cpp
#include <functional>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};


ListNode* plusOne(ListNode* head) {
    std::function<int(ListNode*)> plus = [&](ListNode* n) {
        int res = 1;
        if (n->next) {
            res = plus(n->next);
        }
        n->val += res;
        int ret = n->val / 10;
        n->val %= 10;
        return ret;
    };

    int residue = plus(head);
    if (residue == 1) {
        auto* n = new ListNode(1);
        n->next = head;
        return n;
    }
    return head;
}
